### PR TITLE
test(e2e): extend axe and contrast coverage for showcase homepage

### LINKED ISSUE #684: test(e2e): extend axe and contrast coverage for showcase homepage
## Summary

Make the showcase homepage pass its own "Accessible by default" pitch: axe checks across a sample of themes, keyboard operability of the new interactive elements, and a contrast audit of decorative elements.

## Problem

The existing a11y gate covers the page generically. The showcase adds preview tabs, marquee theme buttons, spotlight/glare decoration, and per-theme rendering — none specifically asserted.

## Proposed Solution

- Extend `e2e/accessibility.spec.ts`: axe scan of the homepage under a representative theme sample (light + dark + high-saturation)
- Keyboard tests: tab order reaches preview tabs and marquee buttons; tabs operable with arrow keys/Enter per ARIA tablist pattern
- Contrast audit of new decorative text against theme backgrounds; fix violations found

## Implementation Notes

- Blocked by #683 (needs the live DOM)
- Reduced-motion behavior is already gated; assert it stays that way

## Benefits

- The a11y required check now defends the redesign's specific surfaces

### PR DESCRIPTION
**Pull request by TurboCoder13**
## Summary

Makes the showcase homepage pass its own "Accessible by default" pitch (#684): the a11y suite now runs axe scans of `/` under a representative theme sample, audits the redesign's decorative text for WCAG AA contrast, and asserts keyboard operability of the new interactive elements. The audit surfaced real violations, fixed in `assets/css/home.css`, and one behavior gap — the preview tabs had no ARIA tablist keyboard support — fixed in `assets/js/homepage-showcase.ts`.

## Type

- [x] ✅ Tests (`test:`)
- [x] 🐛 Bug fix (`fix:`)

### New e2e assertions (`e2e/accessibility.spec.ts`)

- **Per-theme audit** under `catppuccin-latte`, `catppuccin-mocha`, `dracula`, `github-light` (light + dark + high-saturation): full-page axe scan with the suite's standard rule set (zero violations), plus an axe `color-contrast` scan scoped to `.showcase-page` (zero violations)
- **Targeted contrast assertions** for decorative text axe skips or approximates — the hero badge, marquee theme names (inside `aria-label`led buttons), and feature card title/description — via a new `getContrastRatio` e2e helper that alpha-composites translucent `color-mix` backgrounds up the ancestor chain and computes the WCAG ratio (≥ 4.5:1 asserted per theme)
- **Keyboard operability**: tab order reaches the preview tabs and continues through the active panel into the marquee theme buttons; inactive tabs are out of the tab order (roving tabindex); preview tabs follow the ARIA tablist keyboard pattern (ArrowLeft/ArrowRight with wrap-around, Home/End, Enter/Space activation); Enter on a focused marquee card applies the theme through the selector pipeline
- **Reduced motion**: with `reducedMotion: 'reduce'` emulated, the marquee rows and hero reveal report `animation-name: none` and scroll-reveal cards are immediately visible
- Theme CSS is served through a shared `serveThemeCssWithoutExternalImports` helper (extracted from `homepage-theme.spec.ts`'s route-interception pattern) so lazy/theme stylesheet loads stay deterministic under CI's egress policy that blocks external font hosts

### Contrast violations found and fixed (`assets/css/home.css`)

- `.showcase-theme-more` ("View all N themes") was pinned to the Noir chrome's fixed mauve by the sitewide `html:root a` rule — 2.03:1 on github-light, 1.79:1 on catppuccin-latte. Rescoped under `.showcase-page` to out-rank that rule and colored with the theme brand accent mixed 60/40 toward `--turbo-text-primary`
- `.showcase-footer-cta code` used the raw brand accent — 3.71:1 on catppuccin-latte, 3.79:1 on dracula. Same brand/text `color-mix` treatment
- `.showcase-feature h3` inherited per-level theme heading colors that fail AA at 15px — 2.77:1 on catppuccin-latte. Now `--turbo-text-primary`
- `.showcase-feature p` (3.02:1 on dracula, 4.36:1 on catppuccin-latte) and `.showcase-badge` used `--turbo-text-secondary`; nudged toward `--turbo-text-primary` via `color-mix` (60% and 50% secondary respectively), keeping the muted look while clearing 4.5:1 on every sampled theme

### Preview tab keyboard support (`assets/js/homepage-showcase.ts`, `index.astro`)

- APG horizontal tablist pattern: roving tabindex (server-rendered initial state, maintained by `applyTabSelection` and on focus movement), ArrowLeft/ArrowRight with wrap-around, Home/End; activation stays manual via the tabs' native button Enter/Space click
- New pure helpers `nextTabFocusIndex` / `focusTab` exported and unit tested; committed `homepage-showcase` bundles rebuilt with `bun run build:js`

## Closes

- Closes #684

## Testing

- [x] Unit tests added/updated
- [x] Integration tests added/updated
- [x] Manual testing performed
- [x] All tests pass locally (`npm test`)

### Test Coverage

- [x] Coverage maintained or improved
- [x] No decrease in code coverage
- [x] Coverage report reviewed

## Quality Checks

- [x] Linting passes (`npm run lint`)
- [x] Formatting passes (`npm run format`)
- [x] TypeScript build successful (`npm run build`)
- [x] No new warnings or errors
- [x] Markdown linting passes (if applicable)

## Documentation

- [x] Comments added to complex code

## Breaking Changes

- [ ] This PR contains breaking changes

## Deployment Notes

None — CSS-only visual deltas stayed within the visual-regression tolerances (full local run green, no baseline changes needed).

## Checklist

- [x] My code follows the project's style guidelines
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings or errors
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published
- [x] I have checked my code and corrected any misspellings
- [x] **I agree to abide by the project's [Code of Conduct](../CODE_OF_CONDUCT.md)**

## Additional Context

Local gate results: `uv run lintro chk` — 0 issues; `bun run test` — 96 files, 3270 tests passed, coverage thresholds met; full local e2e (`bun run e2e:prep` + full Playwright run, chromium + mobile projects) — 137 passed, 21 skipped (platform skips), 0 failed, including the extended accessibility suite (14/14) and unchanged visual baselines.

## Summary by CodeRabbit

- **Accessibility**
  - Improved homepage showcase tab navigation for keyboard users, including Arrow, Home, and End key support.
  - Ensured only the active preview tab is included in the keyboard tab order.
  - Improved text and link color contrast across showcase elements to better meet WCAG AA guidelines.
  - Preserved reduced-motion behavior by disabling showcase animations when requested.

- **Bug Fixes**
  - Improved keyboard activation and focus behavior for preview themes and showcase controls.

<details><summary><h3>Greptile Summary</h3></summary>

This PR extends the accessibility e2e suite with per-theme axe scans, targeted WCAG AA contrast checks, and keyboard operability assertions for the showcase homepage, and fixes the contrast violations the audit uncovered in `home.css` and the missing ARIA tablist keyboard pattern in `homepage-showcase.ts`.

- **Contrast fixes (`home.css`)**: Four selectors are changed from raw brand or secondary tokens to `color-mix(in srgb, …)` blends, scoping `.showcase-theme-more` under `.showcase-page` to out-rank the sitewide anchor rule.
- **Keyboard support (`homepage-showcase.ts`)**: New `nextTabFocusIndex` / `focusTab` pure helpers implement the APG horizontal tablist pattern (ArrowLeft/Right with wrap, Home/End); `applyTabSelection` now also manages roving tabindex; the initial `tabindex` attribute is server-rendered in `index.astro`.
- **New e2e tests (`e2e/accessibility.spec.ts`)**: Per-theme axe + `color-contrast` scans under four representative themes, targeted `getContrastRatio` checks for decorative text, and full keyboard-operability assertions for the preview tablist and marquee cards; the `serveThemeCssWithoutExternalImports` helper is extracted to shared `e2e/helpers.ts`.
</details>

<h3>Confidence Score: 4/5</h3>

Safe to merge; the CSS contrast fixes, roving-tabindex implementation, and new e2e coverage are all correct and well-tested.

The core logic is correct and well covered by unit and e2e tests. The `getContrastRatio` helper's `parseColor` silently skips unsupported color formats, which could give wrong contrast ratios if a future theme uses a wide-gamut color space. The keyboard tab-order assertion bakes in an implicit assumption about the overview panel's focusable element count. Neither concern affects the correctness of shipped code.

`e2e/helpers.ts` (`getContrastRatio` / `parseColor`) and `e2e/accessibility.spec.ts` (tab-order and route-cleanup patterns) are the areas worth a second look.

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| e2e/accessibility.spec.ts | New per-theme axe + contrast scans and keyboard operability tests; well-structured, but the tab-order assertion makes an implicit assumption about page structure that could silently break. |
| e2e/helpers.ts | Adds `getContrastRatio`, `waitForThemeApplied`, and `serveThemeCssWithoutExternalImports`; the contrast helper's `parseColor` only covers legacy rgb/hsl and `color(srgb …)`, silently skipping other modern color formats. |
| assets/js/homepage-showcase.ts | Adds `nextTabFocusIndex` / `focusTab` pure helpers and wires APG tablist keyboard pattern; logic is correct and well unit-tested. |
| assets/css/home.css | Four contrast fixes using `color-mix(in srgb, …)` blends; specificity scoping for `.showcase-theme-more` is well-explained and correct. |
| apps/site/src/pages/index.astro | Adds server-rendered `tabindex` on tab buttons for correct roving-tabindex initial state; change is minimal and correct. |
| test/homepage-showcase.test.ts | Adds thorough unit tests for `nextTabFocusIndex` and `focusTab`, including edge cases (empty tablist, out-of-range index, wrap-around); all existing tests preserved. |
| e2e/homepage-theme.spec.ts | Route-interception helper extracted to `e2e/helpers.ts`; no functional changes, only import adjustments. |

</details>

<details><summary><h3>Sequence Diagram</h3></summary>

<a href="#gh-light-mode-only">

```mermaid
%%{init: {'theme': 'neutral'}}%%
sequenceDiagram
    participant Test as e2e Test
    participant PW as Playwright Page
    participant Axe as axe-core
    participant Helper as getContrastRatio
    participant TS as homepage-showcase.ts

    Test->>PW: emulateMedia(reducedMotion:'reduce')
    Test->>PW: "route THEME_CSS_ROUTE (strip @import url)"
    Test->>PW: goto('/')
    Test->>PW: switchToTheme(theme)
    PW->>TS: applyTheme(themeId) [integration API]
    TS->>PW: "lazily injects link#turbo-theme-css"
    Test->>PW: "waitForStylesheetLoad(#turbo-theme-css)"
    Test->>PW: waitForThemeApplied(page, theme)
    Test->>Axe: full-page scan (wcag2a/21a/22a, no contrast)
    Axe-->>Test: violations[]
    Test->>Axe: scoped scan (.showcase-page, color-contrast rule)
    Axe-->>Test: violations[]
    loop decorativeText selectors
        Test->>Helper: getContrastRatio(locator)
        Helper->>PW: locator.evaluate() walk ancestor backgrounds
        PW-->>Helper: composited WCAG ratio
        Helper-->>Test: "ratio >= 4.5"
    end
    Test->>PW: unrouteAll()

    Note over Test,TS: Keyboard operability path
    Test->>PW: focus(home-cta-explore-themes)
    Test->>PW: keyboard.press('Tab')
    PW->>TS: "roving tabindex server-rendered tabindex=0 on active tab"
    Test->>PW: expect(overviewTab).toBeFocused()
    Test->>PW: keyboard.press('ArrowRight')
    PW->>TS: "keydown -> nextTabFocusIndex('ArrowRight', 0, 3) -> 1"
    TS->>PW: focusTab(tabs, 1) tabIndex updates + focus()
    Test->>PW: expect(componentsTab).toBeFocused()
    Test->>PW: keyboard.press('Space')
    PW->>TS: "click -> applyTabSelection(tabs, panels, 'components')"
    TS->>PW: aria-selected / hidden / tabIndex updates
```

</a>
<a href="#gh-dark-mode-only">

```mermaid
%%{init: {'theme': 'base', 'themeVariables': {"darkMode": true, "background": "#0d1117", "primaryColor": "#21262d", "primaryTextColor": "#e6edf3", "primaryBorderColor": "#8b949e", "lineColor": "#8b949e", "textColor": "#e6edf3", "edgeLabelBackground": "#161b22", "actorBkg": "#21262d", "actorBorder": "#8b949e", "actorTextColor": "#e6edf3", "actorLineColor": "#8b949e", "signalColor": "#8b949e", "signalTextColor": "#e6edf3", "noteBkgColor": "#373320", "noteBorderColor": "#d4a72c", "noteTextColor": "#f0e6c0", "labelBoxBkgColor": "#21262d", "labelBoxBorderColor": "#8b949e", "labelTextColor": "#e6edf3", "loopTextColor": "#e6edf3", "activationBkgColor": "#30363d", "activationBorderColor": "#8b949e"}}}%%
sequenceDiagram
    participant Test as e2e Test
    participant PW as Playwright Page
    participant Axe as axe-core
    participant Helper as getContrastRatio
    participant TS as homepage-showcase.ts

    Test->>PW: emulateMedia(reducedMotion:'reduce')
    Test->>PW: "route THEME_CSS_ROUTE (strip @import url)"
    Test->>PW: goto('/')
    Test->>PW: switchToTheme(theme)
    PW->>TS: applyTheme(themeId) [integration API]
    TS->>PW: "lazily injects link#turbo-theme-css"
    Test->>PW: "waitForStylesheetLoad(#turbo-theme-css)"
    Test->>PW: waitForThemeApplied(page, theme)
    Test->>Axe: full-page scan (wcag2a/21a/22a, no contrast)
    Axe-->>Test: violations[]
    Test->>Axe: scoped scan (.showcase-page, color-contrast rule)
    Axe-->>Test: violations[]
    loop decorativeText selectors
        Test->>Helper: getContrastRatio(locator)
        Helper->>PW: locator.evaluate() walk ancestor backgrounds
        PW-->>Helper: composited WCAG ratio
        Helper-->>Test: "ratio >= 4.5"
    end
    Test->>PW: unrouteAll()

    Note over Test,TS: Keyboard operability path
    Test->>PW: focus(home-cta-explore-themes)
    Test->>PW: keyboard.press('Tab')
    PW->>TS: "roving tabindex server-rendered tabindex=0 on active tab"
    Test->>PW: expect(overviewTab).toBeFocused()
    Test->>PW: keyboard.press('ArrowRight')
    PW->>TS: "keydown -> nextTabFocusIndex('ArrowRight', 0, 3) -> 1"
    TS->>PW: focusTab(tabs, 1) tabIndex updates + focus()
    Test->>PW: expect(componentsTab).toBeFocused()
    Test->>PW: keyboard.press('Space')
    PW->>TS: "click -> applyTabSelection(tabs, panels, 'components')"
    TS->>PW: aria-selected / hidden / tabIndex updates
```

</a>
</details>

<a href="https://app.greptile.com/api/ide/cursor?prompt=Fix%20the%20following%203%20code%20review%20issues.%20Work%20through%20them%20one%20at%20a%20time%2C%20proposing%20concise%20fixes.%0A%0A---%0A%0A%23%23%23%20Issue%201%20of%203%0Ae2e%2Fhelpers.ts%3A371-394%0A**%60parseColor%60%20silently%20drops%20unrecognised%20color%20formats**%0A%0AThe%20helper%20covers%20%60rgba%3F%28%29%60%2C%20%60hsla%3F%28%29%60%2C%20%60color%28srgb%20%E2%80%A6%29%60%2C%20and%20hex%2C%20but%20modern%20browsers%20can%20return%20other%20serializations%20for%20computed%20colors%20%E2%80%94%20e.g.%20%60oklch%280.97%200.005%20247%29%60%2C%20%60lab%28%E2%80%A6%29%60%2C%20or%20%60color%28display-p3%20%E2%80%A6%29%60%20%E2%80%94%20particularly%20when%20a%20theme%20custom%20property%20is%20authored%20in%20one%20of%20those%20spaces.%20When%20%60parseColor%60%20returns%20%60null%60%20for%20a%20background%20layer%2C%20the%20layer%20is%20silently%20skipped%20%28line%20413%20%60if%20%28parsed%20%26%26%20parsed.a%20%3E%200%29%60%29%3B%20if%20a%20dominant%20translucent%20layer%20is%20skipped%2C%20the%20resolved%20background%20used%20for%20the%20WCAG%20ratio%20computation%20is%20wrong.%20For%20the%20foreground%20text%20color%20%28line%20432%29%20it%20throws%20with%20a%20descriptive%20message%2C%20which%20at%20least%20makes%20the%20problem%20visible%2C%20but%20a%20silent%20wrong-answer%20on%20contrast%20is%20harder%20to%20catch.%0A%0AThe%20current%20theme%20variables%20resolve%20via%20%60color-mix%28in%20srgb%20%E2%80%A6%29%60%20so%20browsers%20serialize%20them%20to%20%60rgb%28%29%60%20or%20%60color%28srgb%20%E2%80%A6%29%60%20today%2C%20but%20adding%20an%20%60oklch%60-based%20theme%20later%20would%20silently%20corrupt%20the%20contrast%20gate.%0A%0A%23%23%23%20Issue%202%20of%203%0Ae2e%2Faccessibility.spec.ts%3A188-196%0A**Tab-order%20assertion%20makes%20an%20implicit%20page-structure%20assumption**%0A%0AThe%20test%20presses%20Tab%20once%20from%20%60home-preview-code-copy%60%20and%20immediately%20expects%20focus%20to%20be%20on%20a%20%60.showcase-marquee-card%60.%20This%20is%20only%20correct%20because%20the%20overview%20panel%20currently%20has%20no%20other%20focusable%20elements%20after%20the%20copy%20button%20%28the%20primary%2Foutline%20buttons%20live%20in%20the%20hidden%20components%20panel%29.%20Any%20future%20interactive%20element%20added%20to%20the%20overview%20panel%20%E2%80%94%20a%20link%2C%20input%2C%20or%20additional%20button%20%E2%80%94%20between%20the%20copy%20button%20and%20the%20marquee%20section%20would%20silently%20break%20this%20assertion%20without%20a%20clear%20error%20message.%20Asserting%20%60page.locator%28'%3Afocus'%29%60%20also%20makes%20the%20failure%20message%20opaque.%20Capturing%20the%20expected%20element%20before%20the%20Tab%20press%20and%20asserting%20%60toBeFocused%28%29%60%20on%20a%20named%20locator%20would%20both%20make%20intent%20clearer%20and%20give%20a%20better%20failure%20message.%0A%0A%23%23%23%20Issue%203%20of%203%0Ae2e%2Faccessibility.spec.ts%3A104-162%0A**Route%20cleanup%20not%20guarded%20by%20%60afterEach%60%20or%20%60try-finally%60**%0A%0A%60serveThemeCssWithoutExternalImports%60%20is%20called%20at%20line%20111%20and%20the%20matching%20%60unrouteAll%60%20is%20at%20line%20160%20inside%20the%20test%20body.%20If%20any%20assertion%20between%20those%20lines%20throws%20%28e.g.%20an%20axe%20violation%20is%20found%20or%20the%20theme%20fails%20to%20apply%29%2C%20%60unrouteAll%60%20is%20never%20reached.%20Playwright's%20fixture%20system%20creates%20a%20fresh%20browser%20context%20per%20test%2C%20so%20the%20route%20handler%20doesn't%20leak%20across%20tests%2C%20but%20any%20in-flight%20route%20callback%20still%20registered%20at%20teardown%20time%20can%20emit%20console%20errors%20or%20make%20the%20test%20report%20noisier%20than%20expected.%20The%20same%20pattern%20appears%20in%20the%20keyboard%20test%20at%20lines%20254%20and%20269.%20Using%20%60test.afterEach%60%20or%20a%20%60try-finally%60%20block%20to%20call%20%60unrouteAll%60%20makes%20cleanup%20unconditional.%0A%0A&pr=702&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInCursorDark.svg?v=6"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInCursor.svg?v=6">Fix All in Cursor</picture></a> <a href="https://app.greptile.com/ide/claude-code?prompt=Fix%20the%20following%203%20code%20review%20issues.%20Work%20through%20them%20one%20at%20a%20time%2C%20proposing%20concise%20fixes.%0A%0A---%0A%0A%23%23%23%20Issue%201%20of%203%0Ae2e%2Fhelpers.ts%3A371-394%0A**%60parseColor%60%20silently%20drops%20unrecognised%20color%20formats**%0A%0AThe%20helper%20covers%20%60rgba%3F%28%29%60%2C%20%60hsla%3F%28%29%60%2C%20%60color%28srgb%20%E2%80%A6%29%60%2C%20and%20hex%2C%20but%20modern%20browsers%20can%20return%20other%20serializations%20for%20computed%20colors%20%E2%80%94%20e.g.%20%60oklch%280.97%200.005%20247%29%60%2C%20%60lab%28%E2%80%A6%29%60%2C%20or%20%60color%28display-p3%20%E2%80%A6%29%60%20%E2%80%94%20particularly%20when%20a%20theme%20custom%20property%20is%20authored%20in%20one%20of%20those%20spaces.%20When%20%60parseColor%60%20returns%20%60null%60%20for%20a%20background%20layer%2C%20the%20layer%20is%20silently%20skipped%20%28line%20413%20%60if%20%28parsed%20%26%26%20parsed.a%20%3E%200%29%60%29%3B%20if%20a%20dominant%20translucent%20layer%20is%20skipped%2C%20the%20resolved%20background%20used%20for%20the%20WCAG%20ratio%20computation%20is%20wrong.%20For%20the%20foreground%20text%20color%20%28line%20432%29%20it%20throws%20with%20a%20descriptive%20message%2C%20which%20at%20least%20makes%20the%20problem%20visible%2C%20but%20a%20silent%20wrong-answer%20on%20contrast%20is%20harder%20to%20catch.%0A%0AThe%20current%20theme%20variables%20resolve%20via%20%60color-mix%28in%20srgb%20%E2%80%A6%29%60%20so%20browsers%20serialize%20them%20to%20%60rgb%28%29%60%20or%20%60color%28srgb%20%E2%80%A6%29%60%20today%2C%20but%20adding%20an%20%60oklch%60-based%20theme%20later%20would%20silently%20corrupt%20the%20contrast%20gate.%0A%0A%23%23%23%20Issue%202%20of%203%0Ae2e%2Faccessibility.spec.ts%3A188-196%0A**Tab-order%20assertion%20makes%20an%20implicit%20page-structure%20assumption**%0A%0AThe%20test%20presses%20Tab%20once%20from%20%60home-preview-code-copy%60%20and%20immediately%20expects%20focus%20to%20be%20on%20a%20%60.showcase-marquee-card%60.%20This%20is%20only%20correct%20because%20the%20overview%20panel%20currently%20has%20no%20other%20focusable%20elements%20after%20the%20copy%20button%20%28the%20primary%2Foutline%20buttons%20live%20in%20the%20hidden%20components%20panel%29.%20Any%20future%20interactive%20element%20added%20to%20the%20overview%20panel%20%E2%80%94%20a%20link%2C%20input%2C%20or%20additional%20button%20%E2%80%94%20between%20the%20copy%20button%20and%20the%20marquee%20section%20would%20silently%20break%20this%20assertion%20without%20a%20clear%20error%20message.%20Asserting%20%60page.locator%28'%3Afocus'%29%60%20also%20makes%20the%20failure%20message%20opaque.%20Capturing%20the%20expected%20element%20before%20the%20Tab%20press%20and%20asserting%20%60toBeFocused%28%29%60%20on%20a%20named%20locator%20would%20both%20make%20intent%20clearer%20and%20give%20a%20better%20failure%20message.%0A%0A%23%23%23%20Issue%203%20of%203%0Ae2e%2Faccessibility.spec.ts%3A104-162%0A**Route%20cleanup%20not%20guarded%20by%20%60afterEach%60%20or%20%60try-finally%60**%0A%0A%60serveThemeCssWithoutExternalImports%60%20is%20called%20at%20line%20111%20and%20the%20matching%20%60unrouteAll%60%20is%20at%20line%20160%20inside%20the%20test%20body.%20If%20any%20assertion%20between%20those%20lines%20throws%20%28e.g.%20an%20axe%20violation%20is%20found%20or%20the%20theme%20fails%20to%20apply%29%2C%20%60unrouteAll%60%20is%20never%20reached.%20Playwright's%20fixture%20system%20creates%20a%20fresh%20browser%20context%20per%20test%2C%20so%20the%20route%20handler%20doesn't%20leak%20across%20tests%2C%20but%20any%20in-flight%20route%20callback%20still%20registered%20at%20teardown%20time%20can%20emit%20console%20errors%20or%20make%20the%20test%20report%20noisier%20than%20expected.%20The%20same%20pattern%20appears%20in%20the%20keyboard%20test%20at%20lines%20254%20and%20269.%20Using%20%60test.afterEach%60%20or%20a%20%60try-finally%60%20block%20to%20call%20%60unrouteAll%60%20makes%20cleanup%20unconditional.%0A%0A&repo=lgtm-hq%2Fturbo-themes&pr=702&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInClaudeDark.svg?v=6"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInClaude.svg?v=6">Fix All in Claude Code</picture></a> <a href="https://app.greptile.com/api/ide/codex?prompt=IMPORTANT%3A%20Work%20in%20the%20repository%20%22lgtm-hq%2Fturbo-themes%22%20on%20the%20existing%20branch%20%22test%2F684-showcase-axe-contrast%22.%20Checkout%20that%20branch%20%E2%80%94%20do%20NOT%20create%20a%20new%20branch%20or%20open%20a%20new%20PR.%20Push%20your%20changes%20to%20%22test%2F684-showcase-axe-contrast%22.%0A%0AFix%20the%20following%203%20code%20review%20issues.%20Work%20through%20them%20one%20at%20a%20time%2C%20proposing%20concise%20fixes.%0A%0A---%0A%0A%23%23%23%20Issue%201%20of%203%0Ae2e%2Fhelpers.ts%3A371-394%0A**%60parseColor%60%20silently%20drops%20unrecognised%20color%20formats**%0A%0AThe%20helper%20covers%20%60rgba%3F%28%29%60%2C%20%60hsla%3F%28%29%60%2C%20%60color%28srgb%20%E2%80%A6%29%60%2C%20and%20hex%2C%20but%20modern%20browsers%20can%20return%20other%20serializations%20for%20computed%20colors%20%E2%80%94%20e.g.%20%60oklch%280.97%200.005%20247%29%60%2C%20%60lab%28%E2%80%A6%29%60%2C%20or%20%60color%28display-p3%20%E2%80%A6%29%60%20%E2%80%94%20particularly%20when%20a%20theme%20custom%20property%20is%20authored%20in%20one%20of%20those%20spaces.%20When%20%60parseColor%60%20returns%20%60null%60%20for%20a%20background%20layer%2C%20the%20layer%20is%20silently%20skipped%20%28line%20413%20%60if%20%28parsed%20%26%26%20parsed.a%20%3E%200%29%60%29%3B%20if%20a%20dominant%20translucent%20layer%20is%20skipped%2C%20the%20resolved%20background%20used%20for%20the%20WCAG%20ratio%20computation%20is%20wrong.%20For%20the%20foreground%20text%20color%20%28line%20432%29%20it%20throws%20with%20a%20descriptive%20message%2C%20which%20at%20least%20makes%20the%20problem%20visible%2C%20but%20a%20silent%20wrong-answer%20on%20contrast%20is%20harder%20to%20catch.%0A%0AThe%20current%20theme%20variables%20resolve%20via%20%60color-mix%28in%20srgb%20%E2%80%A6%29%60%20so%20browsers%20serialize%20them%20to%20%60rgb%28%29%60%20or%20%60color%28srgb%20%E2%80%A6%29%60%20today%2C%20but%20adding%20an%20%60oklch%60-based%20theme%20later%20would%20silently%20corrupt%20the%20contrast%20gate.%0A%0A%23%23%23%20Issue%202%20of%203%0Ae2e%2Faccessibility.spec.ts%3A188-196%0A**Tab-order%20assertion%20makes%20an%20implicit%20page-structure%20assumption**%0A%0AThe%20test%20presses%20Tab%20once%20from%20%60home-preview-code-copy%60%20and%20immediately%20expects%20focus%20to%20be%20on%20a%20%60.showcase-marquee-card%60.%20This%20is%20only%20correct%20because%20the%20overview%20panel%20currently%20has%20no%20other%20focusable%20elements%20after%20the%20copy%20button%20%28the%20primary%2Foutline%20buttons%20live%20in%20the%20hidden%20components%20panel%29.%20Any%20future%20interactive%20element%20added%20to%20the%20overview%20panel%20%E2%80%94%20a%20link%2C%20input%2C%20or%20additional%20button%20%E2%80%94%20between%20the%20copy%20button%20and%20the%20marquee%20section%20would%20silently%20break%20this%20assertion%20without%20a%20clear%20error%20message.%20Asserting%20%60page.locator%28'%3Afocus'%29%60%20also%20makes%20the%20failure%20message%20opaque.%20Capturing%20the%20expected%20element%20before%20the%20Tab%20press%20and%20asserting%20%60toBeFocused%28%29%60%20on%20a%20named%20locator%20would%20both%20make%20intent%20clearer%20and%20give%20a%20better%20failure%20message.%0A%0A%23%23%23%20Issue%203%20of%203%0Ae2e%2Faccessibility.spec.ts%3A104-162%0A**Route%20cleanup%20not%20guarded%20by%20%60afterEach%60%20or%20%60try-finally%60**%0A%0A%60serveThemeCssWithoutExternalImports%60%20is%20called%20at%20line%20111%20and%20the%20matching%20%60unrouteAll%60%20is%20at%20line%20160%20inside%20the%20test%20body.%20If%20any%20assertion%20between%20those%20lines%20throws%20%28e.g.%20an%20axe%20violation%20is%20found%20or%20the%20theme%20fails%20to%20apply%29%2C%20%60unrouteAll%60%20is%20never%20reached.%20Playwright's%20fixture%20system%20creates%20a%20fresh%20browser%20context%20per%20test%2C%20so%20the%20route%20handler%20doesn't%20leak%20across%20tests%2C%20but%20any%20in-flight%20route%20callback%20still%20registered%20at%20teardown%20time%20can%20emit%20console%20errors%20or%20make%20the%20test%20report%20noisier%20than%20expected.%20The%20same%20pattern%20appears%20in%20the%20keyboard%20test%20at%20lines%20254%20and%20269.%20Using%20%60test.afterEach%60%20or%20a%20%60try-finally%60%20block%20to%20call%20%60unrouteAll%60%20makes%20cleanup%20unconditional.%0A%0A&repo=lgtm-hq%2Fturbo-themes&pr=702&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInCodexDark.svg?v=6"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInCodex.svg?v=6">Fix All in Codex</picture></a>

<sub>Reviews (1): Last reviewed commit: ["test(e2e): extend axe and keyboard cover..."](https://github.com/lgtm-hq/turbo-themes/commit/987ddc040ca7fa22705c5b36afd6b87f053c27dc) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=45682820)</sub>

> Greptile also left **3 inline comments** on this PR.

### CODE REVIEW

File: e2e/helpers.ts [hunk 1]
```diff
@@ -330,6 +330,125 @@ export async function removeThemeCSSInterception(page: Page): Promise<void> {
...
+ *
+ * @param page - The Playwright page object
+ */
+export async function serveThemeCssWithoutExternalImports(page: Page): Promise<void> {
+  await page.route(THEME_CSS_ROUTE, async (route) => {
+    const response = await route.fetch();
+    const body = (await response.text()).replace(/@import url\((?:'|")https?:[^)]*\);/g, '');
+    await route.fulfill({ response, body });
+  });
+}
```
[the remark this answers is not in the document]
Comment (TurboCoder13) [reply e2e/helpers.ts:354]: Valid quick win — the helper should drop the original content-length and let route.fulfill recompute it for the rewritten body. Works today (CI green with the same pattern from #701), so deferred to #696 rather than another CI cycle here.

File: e2e/helpers.ts [hunk 2]
```diff
@@ -330,6 +330,125 @@ export async function removeThemeCSSInterception(page: Page): Promise<void> {
...
+      );
+      if (modern) {
+        return {
+          r: Number(modern[1]) * 255,
+          g: Number(modern[2]) * 255,
+          b: Number(modern[3]) * 255,
+          a: modern[4] === undefined ? 1 : Number(modern[4]),
+        };
+      }
+      return null;
```
[the remark this answers is not in the document]
Comment (TurboCoder13) [reply e2e/helpers.ts:394]: Fair point — current theme variables serialize to rgb()/color(srgb) as you note, so behavior is correct today, but a silent skip on future color spaces could distort the ratio. Deferred to #696: support additional spaces or throw like the foreground path.

File: e2e/accessibility.spec.ts [hunk 3]
```diff
@@ -85,6 +91,208 @@ test.describe('Accessibility Tests @a11y', () => {
...
+
+      await test.step('Tab continues into the active panel, then the marquee', async () => {
+        await page.keyboard.press('Tab');
+        await expect(page.getByTestId('home-preview-code-copy')).toBeFocused();
+
+        await page.keyboard.press('Tab');
+        const focused = page.locator(':focus');
+        await expect(focused).toHaveClass(/showcase-marquee-card/);
+        await expect(focused).toHaveAttribute('data-theme-preview', /.+/);
+      });
```
[the remark this answers is not in the document]
Comment (TurboCoder13) [reply e2e/accessibility.spec.ts:196]: Agreed on robustness — asserting toBeFocused() on a captured locator is clearer and fails better. Test is correct against the current DOM; improvement deferred to #696.

File: e2e/accessibility.spec.ts [hunk 4]
```diff
@@ -85,6 +91,208 @@ test.describe('Accessibility Tests @a11y', () => {
...
+            expect(
+              ratio,
+              `${label} contrast ratio under ${theme} (got ${ratio.toFixed(2)})`
+            ).toBeGreaterThanOrEqual(MIN_CONTRAST_NORMAL_TEXT);
+          }
+        });
+
+        await homePage.page.unrouteAll({ behavior: 'ignoreErrors' });
+      });
+    });
```
[the remark this answers is not in the document]
Comment (TurboCoder13) [reply e2e/accessibility.spec.ts:162]: Valid hygiene — contexts are per-test so nothing leaks across tests, but unconditional cleanup via try-finally/afterEach is right. Deferred to #696 with the other e2e items.
